### PR TITLE
DataFrame <-> NumpyEvent

### PR DESCRIPTION
**Pull request by diegomarvid**
- [x] DataFrame To Numpy Event
- [x] Numpy Event To DataFrame
- [x] Tests
- [x] Ensure it works with index with only timestamps

### CODE REVIEW

File: temporian/implementation/numpy/data/event.py [hunk 1]
```diff
@@ -94,6 +103,105 @@ def schema(self) -> Event:
...
+        """
+        df_index = self.sampling.names + [timestamp_index_name]
+        df_features = self.feature_names
+        columns = df_index + df_features
+
+        df = pd.DataFrame(data=[], columns=columns).set_index(df_index)
+
+        for index, features in self.data.items():
+            timestamps_index = self.sampling.data[index]
+            for i, timestamp in enumerate(timestamps_index):
```
Comment (getchebarne) [inline temporian/implementation/numpy/data/event.py:189]: I think we should be able to circumvent this `for` loop. Pandas allows for multiple multi-index assignments. Should be more efficient.

File: temporian/implementation/numpy/data/event.py [hunk 2]
```diff
@@ -94,6 +103,105 @@ def schema(self) -> Event:
...
+                # add timestamp to index
+                new_index = index + (timestamp,)
+                df.loc[new_index, df_features] = [
+                    feature.data[i] for feature in features
+                ]
+
+        # Convert to original dtypes, can be more efficient
+        first_index = self.first_index
+        first_features = self.data[first_index]
```
Comment (getchebarne) [inline temporian/implementation/numpy/data/event.py:198]: how about a `.head` method?

File: temporian/implementation/numpy/data/event.py [hunk 3]
```diff
@@ -94,6 +103,105 @@ def schema(self) -> Event:
...
+        df = df.astype(
+            {feature.name: type(feature.data[0]) for feature in first_features}
```
Comment (getchebarne) [inline temporian/implementation/numpy/data/event.py:200]: Is there any benefit compared to using `feature.data.dtype`?

File: temporian/test/event_conversion_test.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        )
+
+        expected_df = pd.DataFrame(
+            [
+                [666964, 1.0, 740.0],
+                [666964, 2.0, 508.0],
+                [574016, 3.0, 573.0],
+            ],
+            columns=["product_id", "timestamp", "costs"],
+        ).set_index(["product_id", "timestamp"])
```
Comment (getchebarne) [inline temporian/test/event_conversion_test.py:87]: we shouldn't use floats as indexes, we should parse to datetimes first.

**Review by getchebarne — CHANGES_REQUESTED** [review]

File: temporian/implementation/numpy/data/event.py [hunk 5]
```diff
@@ -94,6 +103,105 @@ def schema(self) -> Event:
             sampling=self.sampling.names,
         )
 
+    @staticmethod
+    def from_dataframe(
```
[the remark this answers is not in the document]
Comment (diegomarvid) [reply temporian/implementation/numpy/data/event.py:107]: Yeah, just implemented this

[the remark this answers is not in the document]
Comment (diegomarvid) [reply temporian/implementation/numpy/data/event.py:107]: Looks better btw!

File: temporian/implementation/numpy/data/event.py [hunk 6]
```diff
@@ -89,6 +99,110 @@ def schema(self) -> Event:
...
+                NumpyFeature(feature, df[feature].to_numpy())
+                for feature in feature_columns
+            ]
+
+        numpy_sampling = NumpySampling(index=index_names, data=sampling)
+
+        return NumpyEvent(data=data, sampling=numpy_sampling)
+
+    def to_dataframe(
+        self, timestamp_index_name: str = "timestamp"
```
Comment (achoum) [inline temporian/implementation/numpy/data/event.py:166]: This is a very long parameter name.
What about "timestamp_column"?

File: temporian/implementation/numpy/data/event.py [hunk 7]
```diff
@@ -89,6 +99,110 @@ def schema(self) -> Event:
...
+    ) -> pd.DataFrame:
+        """Function to convert a NumpyEvent to a pandas DataFrame
+
+        Args:
+            timestamp_index_name: Name for timestamp index. Defaults to "timestamp".
+
+        Returns:
+            pd.DataFrame: DataFrame created from NumpyEvent
+        """
+        df_index = self.sampling.index + [timestamp_index_name]
```
Comment (achoum) [inline temporian/implementation/numpy/data/event.py:176]: Can you make sure the timestamp name does not overlap with an index name?
In addition, can you also make sure that no index name overlap with a feature name.

File: temporian/implementation/numpy/data/event.py [hunk 8]
```diff
@@ -89,6 +99,110 @@ def schema(self) -> Event:
...
+        df_features = self.feature_names
+        columns = df_index + df_features
+
+        df = pd.DataFrame(data=[], columns=columns).set_index(df_index)
```
Comment (achoum) [inline temporian/implementation/numpy/data/event.py:180]: I would make the output dataframe have exactly the same format as the expected format in from_dataframe. Like for from_dataframe, I think having the dataframe be without index might be better.

File: temporian/test/event_conversion_test.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+                        data=np.array([740.0, 508.0, 573.0]), name="costs"
+                    ),
+                ],
+            },
+            sampling=numpy_sampling,
+        )
+
+        expected_df = pd.DataFrame(
+            [
+                [666964, 1.0, 740.0],
```
Comment (achoum) [inline temporian/test/event_conversion_test.py:166]: Idea of things to test:
- when the timestamp is a Date.
- when one of the feature is a string.
- when one of the feature contains missing values .

**Review by achoum — COMMENTED** [review]
This is great.
Left a few comments.

File: temporian/implementation/numpy/data/event.py [hunk 10]
```diff
@@ -89,6 +99,110 @@ def schema(self) -> Event:
             sampling=self.sampling.names,
         )
 
+    @staticmethod
+    def from_dataframe(
+        df: pd.DataFrame,
+        index_names: List[str],
+        timestamp_name: str = "timestamp",
```
[the remark this answers is not in the document]
Comment (diegomarvid) [reply temporian/implementation/numpy/data/event.py:106]: I think every pandas dtype should be allowed right? I don't include any constraints in the code.  What do you think about?

File: temporian/test/event_conversion_test.py (hunk 9 above)
Comment (diegomarvid) [reply temporian/test/event_conversion_test.py]: Should we support Date format in timestamp even if it's not in `DTYPE_MAPPING`?

**Comment by achoum** [comment]
LGTM once what we discussed (e.g. timestamps is improvented).

**Review by achoum — APPROVED** [review]

File: temporian/implementation/numpy/data/event.py [hunk 11]
```diff
@@ -89,6 +101,166 @@ def schema(self) -> Event:
...
+                NumpyFeature(feature, df[feature].to_numpy())
+                for feature in feature_columns
+            ]
+
+        numpy_sampling = NumpySampling(index=index_names, data=sampling)
+
+        return NumpyEvent(data=data, sampling=numpy_sampling)
+
+    def to_dataframe(self) -> pd.DataFrame:
+        """Function to convert a NumpyEvent to a pandas DataFrame
```
Comment (ianspektor) [inline temporian/implementation/numpy/data/event.py:217]: nit: missing dots at the end of most sentences in docstrings

File: temporian/test/event_conversion_test.py [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+            sampling=numpy_sampling,
+        )
+
+        numpy_event = NumpyEvent.from_dataframe(
+            df, index_names=["product_id"], timestamp_column="timestamp"
+        )
+
+        # validate
+        self.assertEqual(
+            True,
```
Comment (ianspektor) [inline temporian/test/event_conversion_test.py:59]: `self.assertTrue`

File: temporian/implementation/numpy/data/event.py [hunk 13]
```diff
@@ -89,6 +105,165 @@ def schema(self) -> Event:
...
+        timestamp_column: str = "timestamp",
+    ) -> "NumpyEvent":
+        """Convert a pandas DataFrame to a NumpyEvent. Supported
+        dtypes are: np.float64, np.float32, np.int64, np.int32, np.datetime64
+        (for timestamp column).
+
+        Args:
+            df: DataFrame to convert to NumpyEvent.
+            index_names: names of the DataFrame columns to be used as index for the event.
+            timestamp_column: Name for timestamp index. Defaults to "timestamp".
```
Comment (achoum) [inline temporian/implementation/numpy/data/event.py:121]: Column containing timestamps. Supported types: {...}. Timestamps of the types [...] are converted implicitly to in UTC epoch.

File: temporian/implementation/numpy/data/event.py [hunk 14]
```diff
@@ -89,6 +105,165 @@ def schema(self) -> Event:
             sampling=self.sampling.names,
         )
 
+    @staticmethod
+    def from_dataframe(
+        df: pd.DataFrame,
+        index_names: List[str],
```
Comment (achoum) [inline temporian/implementation/numpy/data/event.py:111]: Having the indexing logic is great. However, would it make sense to move this logic to the "addIndex" (or something similar) operator instead?

To speed-up the work, you can submit this PR as it is, and maybe make the move in another PR.

File: temporian/implementation/numpy/data/event.py [hunk 15]
```diff
@@ -89,6 +105,165 @@ def schema(self) -> Event:
...
+
+        # check timestamp_column is not on index_names
+        if timestamp_column in index_names:
+            raise ValueError(
+                f"Timestamp column {timestamp_column} cannot be on index_names"
+            )
+
+        # check column dtypes, every dtype should be a key of DTYPE_MAPPING
+        for column in df.columns:
+            # if it's a date, convert it to duration
```
Comment (achoum) [inline temporian/implementation/numpy/data/event.py:164]: Nit: s/duration/UTC epoch

File: temporian/implementation/numpy/data/event.py [hunk 16]
```diff
@@ -89,6 +105,165 @@ def schema(self) -> Event:
...
+            if is_a_date(df[column].dtype.type):
```
Comment (achoum) [inline temporian/implementation/numpy/data/event.py:165]: I would only do an automated conversion for the timestamp column.

File: temporian/implementation/numpy/data/tests/test_df_to_event.py [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+            df, index_names=["product_id"], timestamp_column="timestamp"
+        )
+
+        # validate
+        self.assertTrue(numpy_event == expected_numpy_event)
+
+    def test_df_to_numpy_event_datetime_index(self) -> None:
+        df = pd.DataFrame(
+            [
+                [666964, np.datetime64("2022-01-01"), 740.0],
```
Comment (achoum) [inline temporian/implementation/numpy/data/tests/test_df_to_event.py:115]: Can you test the other supported datetime object, and ensure that the timestamp is as expected?

**Review by achoum — APPROVED** [review]

File: temporian/implementation/numpy/data/event.py (hunk 16 above)
Comment (diegomarvid) [reply temporian/implementation/numpy/data/event.py:165]: And what would happen for other columns if the DTYPE is a date? Should we raise an error?

File: temporian/core/data/duration.py [hunk 18]
```diff
@@ -36,3 +41,65 @@ def hours(value: float) -> Duration:
...
+
+def convert_struct_time_to_duration(date: time.struct_time) -> Duration:
+    """Convert struct_time to duration epoch UTC"""
+    return float(time.mktime(date) - time.timezone)
+
+
+def convert_datetime_to_duration(date: datetime.datetime) -> Duration:
+    """Convert datetime to duration epoch UTC"""
+    date = date.replace(tzinfo=datetime.timezone.utc)
+    return float(date.timestamp())
```
[the remark this answers is not in the document]
Comment (diegomarvid) [reply temporian/core/data/duration.py]: If it was a float, it would return false in the `is_a_date` function and therefore no conversion will be needed.  This only converts datetime objects.
